File: medium/0636_exclusive_time_of_functions/solution.cpp
using namespace std;

#include <stack>
#include <string>
#include <vector>

class Solution {
public:
  vector<int> exclusiveTime(int n, vector<string> &logs) {
    stack<int> s;
    vector<int> times(n, 0);

    int lastTs;

    for (const string &log : logs) {
      int fid = stoi(log.substr(0, log.find(':')));
      char action = log.find('s') != -1 ? 's' : 'e';
      int ts = stoi(log.substr(log.rfind(':') + 1));

      if (action == 's') {
        if (!s.empty())
          times[s.top()] += ts - lastTs;

        s.push(fid);
        lastTs = ts;
      } else {
        times[s.top()] += ts - lastTs + 1;
        s.pop();

        if (!s.empty())
          lastTs = ts + 1;
      }
    }

    return times;
  }
};

File: medium/0636_exclusive_time_of_functions/solution.py
from typing import List


class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        stack = []
        times = {}

        for log in logs:
            fid, action, ts = log.split(":")

            fid = int(fid)
            ts = int(ts)

            if fid not in times:
                times[fid] = 0

            if action == "start":
                if stack:
                    top = stack[-1]
                    times[top[0]] += ts - top[1]

                stack.append([fid, ts])
            else:
                fid, start_ts = stack.pop()
                times[fid] += ts - start_ts + 1

                if stack:
                    stack[-1][1] = ts + 1

        return [time for _, time in sorted(times.items())]
